_posts/2018-04-17-implementing-paging-library-with-retrofit-part-1.md

```diff
@@ -0,0 +1,188 @@
+
+<!--more-->
+
+To demonstrate the paging library, we'll build a simple PokeDex app that displays a list of Pokemon, automatically loading more for us as we scroll:
+
+![AndroidEssence](https://thepracticaldev.s3.amazonaws.com/i/am2687bnk1e3vzro9jlg.gif)
+
+## Background
+
+This post will not go over Retrofit directly, but you can view that in a [previous AndroidEssence](https://androidessence.com/android/getting-started-with-retrofit/) post. It will help to familiarize yourself with the [PokeApi](https://pokeapi.co/), specifically calling for a list of pokemon with some limit and offset, which we can do like this: `https://pokeapi.co/api/v2/pokemon/?limit=2&offset=0`. 
+
+Please be aware that at the time of publishing the paging library is also in beta, so the API is subject to changes.
+
+## DataSource
+
+Diving into the paging library, the first class we want to understand is the [DataSource](https://developer.android.com/reference/android/arch/paging/DataSource.html). This class is used to supply data to a [PagedList](https://developer.android.com/reference/android/arch/paging/PagedList.html), which is simply a list implementation that loads pages/chunks of data from the DataSource. 
+
+There are multiple implementations of DataSource, but for this example we want a [PageKeyedDataSource](https://developer.android.com/reference/android/arch/paging/PageKeyedDataSource.html). Notice how in our API call we can paginate through Pokemon by changing the `offset` parameter? Since that offset parameter works like a "page" number for which subset of Pokemon we want, and it's the only thing that changes in each call, PageKeyedDataSource is the one we want. 
+
+Let's start by defining our DataSource, including any properties we will need and the retry mechanism. Note that the generic types are for the pagination key (in this example, an Int), and the type of object in the resulting list (in this example, a BasePokemon):
+
+```kotlin
+    /**
+     * Paged data source for fetching a list of pokemon. A [PageKeyedDataSource] is helpful when the
+     * data is fetched by a key such as a page number, or in this example an offset.
+     */
+    class PageKeyedBasePokemonDataSource(private val pokeAPI: PokeAPI, private val retryExecutor: Executor) : PageKeyedDataSource<Int, BasePokemon>() {
+
+        // Keep a function reference for the retry event
+        private var retry: (() -> Any)? = null
+
+        // One network state is used while we are loading and paging, but the initial load is always
+        // called first so we need to keep track of that state as well.
+        val networkState = MutableLiveData<NetworkState>()
+        val initialLoad = MutableLiveData<NetworkState>()
```

Comment:
Remove this, it's not relevant anymore because we didn't bother to cover swipe to refresh.